metadata-ingestion/src/datahub/cli/ingest_cli.py

```diff
@@ -198,6 +200,146 @@ async def run_func_check_upgrade(pipeline: Pipeline) -> None:
+)
+@click.option(
+    "--urn",
+    type=str,
+    help="Urn of recipe to update",
+    required=False,
+)
+@click.option(
+    "--executor-id",
+    type=str,
+    default="default",
+    help="Executor id to route execution requests to. Do not use this unless you have configured a custom executor.",
+    required=False,
+)
+@click.option(
+    "--cli-version",
+    type=str,
+    help="Provide a custom CLI version to use for ingestion. By default will use server default.",
+    required=False,
+    default=None,
+)
+@click.option(
+    "--schedule",
+    type=str,
+    help="Cron definition for schedule. If none is provided, ingestion recipe will not be scheduled",
+    required=False,
+    default=None,
+)
+@click.option(
+    "--time-zone",
+    type=str,
+    help=f"Timezone for the schedule. By default uses the timezone of the current system: {tzlocal.get_localzone_name()}.",
+    required=False,
+    default=tzlocal.get_localzone_name()
+)
+def deploy(
+    name: str,
+    config: str,
+    urn: str,
+    executor_id: str,
+    cli_version: str,
+    schedule: str,
+    time_zone: str,
+) -> None:
+    """
+    Deploy an ingestion recipe to your DataHub instance.
+
+    The urn of the ingestion source will be based on the name parameter in the format:
+    urn:li:dataHubIngestionSource:<name>
+    """
+
+    datahub_graph = get_default_graph()
+
+    graphql_query_template = Template(
+        """
+        mutation $endpoint ( $variable_types ) {
+            $endpoint ( $input )
+        }""")
+
+    pipeline_config = load_config_file(
+        config,
+        allow_stdin=True,
+        resolve_env_vars=False,
+    )
+
+    variables: dict = {
+        "name": name,
+        "type": pipeline_config["source"]["type"],
+        "recipe": json.dumps(pipeline_config),
+        "executorId": executor_id,
+    }
+    graphql_variable_types: dict = {
+        "$name": "String!",
+        "$type": "String!",
+        "$recipe": "String!",
+        "$executorId": "String!",
+    }
+    input_template: dict = {
+        "input": {
+            "name": "$name",
+            "type": "$type",
+            "config": {
+                "recipe": "$recipe",
+                "executorId": "$executorId"
+            }
+        }
+    }
+    graphql_endpoint: str
+
+    if urn:
+        if not datahub_graph.exists(urn):
+            logger.error(f"Could not find recipe for provided urn: {urn}")
+            exit()
+        logger.info(f"Found recipe URN, will update recipe.")
+        graphql_endpoint = "updateIngestionSource"
+        input_template["urn"] = "$urn"
+        graphql_variable_types["$urn"] = "String!"
+        variables["urn"] = urn
+    else:
+        logger.info(f"No URN specified recipe urn, will create a new recipe.")
+        graphql_endpoint = "createIngestionSource"
+
+    if schedule:
+        input_template["input"]["schedule"] = "$schedule"
+        graphql_variable_types["$schedule"] = "UpdateIngestionSourceScheduleInput"
```

Comment:
with graphql, you have have the query always take a `$schedule` param. If it doesn't have an `!` at the end of the type, you can omit it and it will use null instead

The whole `graphql_variable_types` thing feels a bit too complex - a static graphql query (or possibly two for create/update) should be sufficient

Reply:
I don’t understand what you mean with the first paragraph. Can you clarify?